dali/kernels/slice/slice_flip_normalize_permute_pad_cpu.h

```diff
@@ -30,165 +31,216 @@ namespace kernels {
+  } else {
+    for (int64_t i = 0; i < out_shape[d]; i++, output += out_strides[d], input += in_strides[d])
+      SliceFlipNormalizePermuteKernelImpl<NeedNormalize, HasChannels>(
+          output, input, in_strides + 1, out_strides + 1, anchor + 1, in_shape + 1, out_shape + 1,
+          mean, inv_stddev, channel_dim - 1, std::integral_constant<int, DimsLeft - 1>());
   }
 }
 
-template <typename Policy, bool IsNormalizationDim, typename OutputType, typename InputType,
-          int Dims>
-inline void SliceFlipNormalizePermuteImpl(OutputType *output, const InputType *input,
-                                          TensorShape<Dims> in_strides,
-                                          TensorShape<Dims> out_strides,
-                                          TensorShape<Dims> out_shape,
-                                          TensorShape<Dims> padded_out_shape,
-                                          const float *mean, const float *inv_stddev,
-                                          int normalization_dim,
-                                          float fill_value,
-                                          std::integral_constant<int, 1>) {
-  constexpr auto d = Dims - 1;
-  int64_t i = 0;
-  for (; i < out_shape[d]; i++) {
-    const int norm_idx = IsNormalizationDim ? i : 0;
-    Policy::Fill(*output, *input, mean + norm_idx, inv_stddev + norm_idx);
-    input += in_strides[d];
-    output += out_strides[d];
-  }
+template <bool NeedNormalize, bool HasChannels, bool OutOfBounds, typename OutputType,
+          typename InputType>
+void SliceFlipNormalizePermutePadKernelImpl(
+    OutputType *output, const InputType *input, const int64_t *in_strides,
+    const int64_t *out_strides, const int64_t *anchor, const int64_t *in_shape,
+    const int64_t *out_shape, const OutputType *fill_values, const float *mean,
+    const float *inv_stddev,
+    int channel_dim,  // negative if no channel dim or already processed
+    std::integral_constant<int, 1>) {
+  constexpr int d = 0;
+  // Note: out_strides[d] is 1 so we can just do output++ in the loops
+  if (OutOfBounds) {
+    if (HasChannels && d == channel_dim) {
+      for (int i = 0; i < out_shape[d]; i++)
+        output[i] = *fill_values++;
+    } else {
+      for (int i = 0; i < out_shape[d]; i++)
+        output[i] = *fill_values;
+    }
+  } else  {
+    int64_t pad_before, slice, pad_after;
+    std::tie(pad_before, slice, pad_after) =
+      CalcPadCopyExtents(anchor[d], in_shape[d], out_shape[d]);
+    if (in_strides[d] < 0) std::swap(pad_before, pad_after);
 
-  // pad
-  for (; i < padded_out_shape[d]; i++) {
-    *output = Convert<OutputType>(fill_value);
-    output += out_strides[d];
+    // out of bounds (left side)
+    if (pad_before > 0) {
+      if (HasChannels && d == channel_dim) {
+        for (int64_t i = 0; i < pad_before; i++)
+          *output++ = *fill_values++;
```

Comment:
I would add `{}` around this line as it is easy to forget that this loop executes only this line.

Reply:
done